### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/{ExteriorAlgebra,CliffordAlgebra}): Functoriality of the exterior algebra and some lemmas about generation

### PR DESCRIPTION
**Pull request by smorel394**
This does a few things:
- Define the algebra morphism `TrivSqZeroExt.map f` between trivial square-zero extensions induced by a linear map`f`, and establish some of its basic properties (functoriality, composition with the basic maps to/from `TrivSqZeroExt`). Note that we only consider the case of a commutative base ring, because the case of a general base ring requires morphisms of bimodules, which we do not have.
- Define the algebra morphism `ExteriorAlgebra.map f` between exterior algebras induced by a linear map `f`. This is a straightforward application of the similar construction for Clifford algebras, but I think it is still useful to have. Basic properties of this construction are proved: functoriality, composition with `ExteriorAlgebra.ι`, `ExteriorAlgebra.ιInv` (this part uses the first point) and `ExteriorAlgebra.ιMulti`. Then exactness properties of the construction is studied:
  - If the linear map is surjective, then the map on exterior algebras is also surjective. This actually holds for Clifford algebras, so I added a lemma called `CliffordAlgebra.map_surjective` in `LinearAlgebra/CliffordAlgebra/Basic.lean`. For exterior algebras, the converse holds and is also proved.
  - If the linear map has a retraction, then the map on exterior algebras is injective. So if the base ring is a field, the map on exterior algebras is injective if the linear map is injective.

- Establish some properties of `ExteriorAlgebra.ιMulti`:
  - `ExteriorAlgebra.ιMulti_range`: The range of `ιMulti R n` is contained in the `n`th exterior power (define here as `LinearMap.range (ι R : M →ₗ[R] ExteriorAlgebra R M) ^ n`).
  - `ExteriorAlgebra.ιMulti_span_fixedDegree`: This range spans the `n`th exterior power.
  - `ExteriorAlgebra.ιMulti_span`: The union over all `n` of the range of `ιMulti R n` spans the whole exterior algebra (this is in `LinearAlgebra/ExteriorAlgebra/Grading.lean` because the proof uses the graded module structure, but it might be possible to do something simpler).

- Construct `ExteriorAlgebra.ιMulti_family`: This takes a natural number `n` and a family of vectors `v` indexed by a linearly ordered type `I`, and it returns the family of `n`-fold products of the `v i` in the exterior algebra, indexed by the set of finsets of `I` of cardinality `n`. (The point, to be proved in another PR, is that when `v` is a basis, then `ExteriorAlgebra.ιMulti_family R n v` is a basis of the `n`th exterior power.)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
Sorry, I think I missed this one. Can you resolve the merge conflicts? Are all the changes in the conflicting file already merged?

**Comment by smorel394** [comment]
Sorry, missed your comment (busy week). I'm thinking of just doing a fresh branch and fresh PR at this point, so the conflicts will be easier to solve.

File: Mathlib/LinearAlgebra/CliffordAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/CliffordAlgebra/Basic.lean:355]:
```suggestion
maps it induces between Clifford algebras is also surjective.-/
```

File: Mathlib/LinearAlgebra/CliffordAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/CliffordAlgebra/Basic.lean:358]: I think this is just making work for yourself; can't you work with the definition of `Surjective` directly?

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:367]: This is computable :)

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:369]: This will be more pleasant to unfold as 
```suggestion
  fun s ↦ ιMulti R n (fun i => v (Finset.orderIsoOfFin _ s.prop i))
```
at which point you may as well move the `s` behind the colon.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Grading.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Grading.lean:103]: A little easier to read with fewer parens, perhaps:
```suggestion
    Submodule.span R (Set.range fun x : Σ n, (Fin n → M) => ιMulti R x.1 x.2) = ⊤ := by
```

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Grading.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Grading.lean:118]: Golfed:
```suggestion
  rw [Submodule.eq_top_iff']
  intro x
  induction x
    using DirectSum.Decomposition.inductionOn fun i => LinearMap.range (ι R (M := M)) ^ i with
  | h_zero => exact Submodule.zero_mem _
  | h_add _ _ hm hm' => exact Submodule.add_mem _ hm hm'
  | h_homogeneous hm =>
    let ⟨m, hm⟩ := hm
    apply Set.mem_of_mem_of_subset hm
    rw [← ιMulti_span_fixedDegree]
    refine Submodule.span_mono fun _ hx ↦ ?_
    obtain ⟨y, rfl⟩ := hx
    exact ⟨⟨_, y⟩, rfl⟩
```

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:351]:
```suggestion
    LinearMap.range (ι R : M →ₗ[R] ExteriorAlgebra R M) ^ n := by
```

**Review by eric-wieser — COMMENTED** [review]
This is looking pretty good now! I suspect this is similar to what I was thinking of when I added `MultilinearMap.range`, but I don't think that definition is of any use for this PR.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: Okay, but I ran into trouble with the power, and the only way I found out of it was to coerce to `Submodule`, raise to the `n`th power then coerce to `Set`.

**Comment by eric-wieser** [comment]
>  (The point, to be proved in another PR, is that when v is a basis, then ExteriorAlgebra.ιMulti_family R n v is a basis of the nth exterior power.)

Note that there is a result about the basis of an exterior algebra already at https://github.com/eric-wieser/lftcm2023-clifford_algebra/blob/master/Projects/ExteriorAlgebra/Targets.lean#L74-L79; though that doesn't cluster by degree in the way that you're working towards (and isn't sorry-free!)

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: Does writing `(LinearMap.range (ι R : M →ₗ[R] ExteriorAlgebra R M) ^ n :)` work? (with the 🙂 at the end)

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: No, I get
```
type mismatch
  LinearMap.range (ι R) ^ n
has type
  Submodule R (ExteriorAlgebra R M) : outParam (Type (max u1 u2))
but is expected to have type
  Set (ExteriorAlgebra R ?m.262794) : Type (max u1 ?u.262791)
```

**Comment by smorel394** [comment]
> > (The point, to be proved in another PR, is that when v is a basis, then ExteriorAlgebra.ιMulti_family R n v is a basis of the nth exterior power.)
> 
> Note that there is a result about the basis of an exterior algebra elsewhere at https://github.com/eric-wieser/lftcm2023-clifford_algebra/blob/master/Projects/ExteriorAlgebra/Targets.lean#L74-L79; though that doesn't cluster by degree in the way that you're working towards (and isn't sorry-free!)

Noted. I have the result about the basis of the exterior powers written up (and sorry-free) [there](https://github.com/leanprover-community/mathlib4/tree/ExteriorPower), but maybe it's possible to merge the projects ?

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:429]: Will do tomorrow, I really need to go to bed now.

**Comment by eric-wieser** [comment]
Ah, perhaps I falsely assumed that you cared about the basis of the entire algebra. Indeed, the project I ran at LFTCM2023 only concerned itself with the algebra, not the subspace of powers.

**Comment by smorel394** [comment]
> Ah, perhaps I falsely assumed that you cared about the basis of the entire algebra. Indeed, the project I ran at LFTCM2023 only concerned itself with the algebra, not the subspace of powers.

I care about both, but what I really need for my own project is a basis of each exterior power.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: Wow, black magic. Thanks !

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: I wrote the lemma about left inverses. For the linear maps, I'm keeping the "left inverse" condition as `g.comp f = LinearMap.id` rather than `LeftInverse g f`, because this is the condition that [LinearMap.exists_leftInverse_of_injective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Basis/VectorSpace.html#LinearMap.exists_leftInverse_of_injective) gives me in its conclusion, so it makes the lemma easier to use for modules over a field.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:440]: nit:
```suggestion
/-- An injective morphism of vector spaces induces an injective morphism of exterior algebras. -/
```
(throughout)

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: Like so ?

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:430]: I pushed a change here to turn this into a `Iff`; are any of your other results amenable to the same strengthening?

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:426]: Can you prove this for `CliffordAlgebra` too?

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:426]: Oh, it's not true (in general)!

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:426]: One direction is very easy (and has the same proof as for the exterior algebra). The other direction is true if `[CliffordAlgebra.ι](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/CliffordAlgebra/Basic.html#CliffordAlgebra.%CE%B9)` is injective, which I thought was true but I'm not used to Clifford algebras (especially over general rings), so who knows. 
When I first wrote the injectivity lemma, I decided against having a Clifford algebra version because my goal was "over a field, `map f` is injective if `f` is injective", which is false in general. But it would be quite easy to add at least one direction of the lemma above for Clifford algebras, if you think that has value (I am not sure myself).

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:430]: `ExteriorAlgebra.map_surjective` is an `Iff`, but the best way I see to prove it would to first prove that `ιInv` and `map f` compose to `f`.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:426]: > The other direction is true if [CliffordAlgebra.ι](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/CliffordAlgebra/Basic.html#CliffordAlgebra.%CE%B9) is injective, which I thought was true but I'm not used to Clifford algebras

This is true whenever `Q` is induced by a bilinear form, though mathlib only actually states it (via `equivExterior`) for rings with an invertible 2. I have plans to tidy that up, as the proof would all work in the more general case.

I don't know if it's true if `Q` is not induced by a bilinear form.

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:426]: > I think it's worth having one direction of the result, simply because it shortens the proof here.

Ok, just a second.

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:426]: Done, thought probably not optimal.
As for the question of generalizing to Clifford algebras, I think it might be too ambitious for this PR.

File: Mathlib/LinearAlgebra/CliffordAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/CliffordAlgebra/Basic.lean:358]:
```suggestion
lemma leftInverse_map_of_leftInverse {Q₁ : QuadraticForm R M₁} {Q₂ : QuadraticForm R M₂} (f : Q₁ →qᵢ Q₂)
```

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: That sounds like a nice result to have, but I think it's fair to declare it out of scope for this PR.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:367]: Could eliminte the `fun` here:
```suggestion
def ιMulti_family (n : ℕ) {I : Type*} [LinearOrder I] (v : I → M) (s : {s : Finset I // Finset.card s = n}) :
    ExteriorAlgebra R M :=
  ιMulti R n fun i => v (Finset.orderIsoOfFin _ s.prop i)
```
I'd be tempted to make this an `abbrev` rather than a `def`, especially since you include no theorems about it.
If it's an `abbrev`, no one needs any theorems because it because everything can see through it.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: I'm seeing if I can do an easy proof. If I fail, then I'll agree with you.

**Review by eric-wieser — APPROVED** [review]
I'll wait to hear your conclusion on the Iff version of the Surjective proof, but otherwise this looks good to go modulo one more small comment. Thanks!

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:478]: Okay, it wasn't that bad. I left `mapTriv` and `snd_comp_mapTriv` in the same file for now, but they should probably be moved [there](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/TrivSqZeroExt.html), with better names.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:448]: These two results should move to `TrivSqZeroExt.lean`

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: I moved the def and theorem. Morally these should also be done for a noncommutative base ring, but it would require `f` to be `R`-bilinear, not just `R`-linear, so it might complicate things.

File: Mathlib/Algebra/TrivSqZeroExt.lean
Comment (eric-wieser) [inline Mathlib/Algebra/TrivSqZeroExt.lean:945]: I don't think this should be primed:
```suggestion
def map (f : M →ₗ[R'] N) : TrivSqZeroExt R' M →ₐ[R'] TrivSqZeroExt R' N := by
```

File: Mathlib/Algebra/TrivSqZeroExt.lean
Comment (eric-wieser) [inline Mathlib/Algebra/TrivSqZeroExt.lean:958]: I think you can get here faster with `liftEquivOfComm`, unless I'm missing something.

File: Mathlib/Algebra/TrivSqZeroExt.lean
Comment (eric-wieser) [inline Mathlib/Algebra/TrivSqZeroExt.lean:944]:
```suggestion
/-- Functoriality of `TrivSqZeroExt` when the ring is commutative: a linear map
`f : M →ₗ[R'] N` induces a morphism of `R'`-algebras from `TrivSqZeroExt R' M` to
`TrivSqZeroExt R' N`.

Note that we cannot neatly state the non-commutative case, as we do not have morphisms of bimodules.
-/
```

File: Mathlib/Algebra/TrivSqZeroExt.lean
Comment (smorel394) [reply Mathlib/Algebra/TrivSqZeroExt.lean]: Yes, much faster. I also wrote some `simp` lemmas.

File: Mathlib/Algebra/TrivSqZeroExt.lean
Comment (eric-wieser) [inline Mathlib/Algebra/TrivSqZeroExt.lean:939]: I think it would be reasonable to add `section map` around all of these new results.

File: Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:463]: This statement is backwards with respect to its name.
```suggestion
    Function.Surjective (map f) ↔ Function.Surjective f := by
```

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean:463]: I'm not sure why, but ok.

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: `foo_iff` is how statements of the form `Foo ↔ Bar` are named. Statements of the form `Bar ↔ Foo ` are called `bar_iff` or `bar_iff_foo`.

In this case, putting the `map` on the left also makes this a nice `simp` lemma.

Comment (smorel394) [reply Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean]: Ah, I see, thanks. Then I should actually make it a 'simp' lemma then.

File: Mathlib/Algebra/TrivSqZeroExt.lean
Comment (eric-wieser) [inline Mathlib/Algebra/TrivSqZeroExt.lean:951]: Please mention this in the commit message

**Review by eric-wieser — APPROVED** [review]
maintainer merge

I pushed some more small cleanups, and fixed a porting note that was causing you to need `erw`. This now looks great! Can you check the commit message reflects what we've ended up at?

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
Thanks! Functoriality is wonderful. 

Assuming the profiling comes back ok, I will merge.

File: Mathlib/Algebra/TrivSqZeroExt.lean
Comment (smorel394) [reply Mathlib/Algebra/TrivSqZeroExt.lean:951]: I updated the commit message.

**Comment by mattrobball** [comment]
+100 lines or so for +50% in instructions isn't great. These mainly all come from typeclass search. But the problems surfaced are not particular to this file and need to be fixed more generally.

bors merge

**Comment by eric-wieser** [comment]
@mattrobball, I tried making some `simp`s unfold les, lets see if it helps.

**Comment by eric-wieser** [comment]
!bench

**Comment by eric-wieser** [comment]
Well, I guess .4% is better than nothing!

bors merge

**Comment by mattrobball** [comment]
Overall instructions were up by 4B though

**Comment by smorel394** [comment]
Great, now I can PR the exterior power stuff (after I have checked that it still builds).

**Comment by smorel394** [comment]
Here it is: #10654

**Pull request closed without merging**
